File: app/services/notification_service.py
"""
app/services/notification_service.py — 외부 알림 발송
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
[역할]
  NotifierAgent에서 호출해 Slack 알림을 발송한다.
  subTask별 배정 요약을 메시지에 포함한다.

[알림 전송 실패 처리]
  Slack 웹훅 오류가 전체 워크플로우를 중단시키면 안 된다.
  예외를 던지지 않고 False를 반환하며,
  NotifierAgent.validate()도 항상 True를 반환해 알림 실패를 허용한다.

[SLACK_WEBHOOK_URL 미설정]
  .env에 값이 없으면 전송을 건너뛰고 빈 channels 목록을 반환한다.
  로컬 개발 환경에서는 Slack 없이도 워크플로우 전체가 동작한다.
"""

from typing import List

import httpx

from app.config import config


class NotificationService:

    async def send_slack(self, message: str) -> bool:
        """
        Slack Incoming Webhook으로 메시지를 비동기 전송한다.
        httpx.AsyncClient를 사용해 이벤트 루프를 블로킹하지 않는다.
        """
        if not config.slack_webhook_url:
            print("[NotificationService] Slack webhook not configured, skipping")
            return False
        try:
            async with httpx.AsyncClient() as client:
                resp = await client.post(
                    config.slack_webhook_url,
                    json={"text": message},
                    timeout=5,
                )
                return resp.status_code == 200
        except Exception as e:
            print(f"[NotificationService] Slack error: {e}")
            return False

    async def notify(
        self,
        task_id: str,
        assignment_summaries: List[str],
    ) -> List[str]:
        """
        subTask별 배정 요약을 포함한 워크플로우 완료 알림을 발송한다.
        성공한 채널 목록을 반환한다.
        """
        if assignment_summaries:
            body = "\n".join(assignment_summaries)
        else:
            body = "  • 배정 정보 없음"

        msg = (
            f":white_check_mark: *SMAAS* 업무 배정 완료\n"
            f"• Task: `{task_id}`\n"
            f"• subTask별 담당자:\n{body}"
        )
        channels: List[str] = []
        if await self.send_slack(msg):
            channels.append("slack")
        return channels

File: src/services/NotificationService.ts
/**
 * services/NotificationService.ts
 * Slack, Email, Webhook 등 외부 채널로 알림을 전송하는 서비스.
 * NotifierAgent에서 직접 외부 API를 호출하지 않고 이 서비스에 위임한다.
 * 채널별 포맷 변환과 전송 실패 처리를 담당한다.
 */

import https from "https";
import { AgentContext } from "../types/workflow.types";
import { NotificationResult } from "../types/agent.types";
import { config } from "../config";

export class NotificationService {
  async send(context: AgentContext): Promise<NotificationResult> {
    const channels: string[] = [];
    const errors: string[] = [];

    if (config.slack.webhookUrl) {
      try {
        await this.sendSlack(context);
        channels.push("slack");
      } catch (err) {
        errors.push(`slack: ${(err as Error).message}`);
      }
    }

    if (config.smtp.host && config.smtp.user) {
      try {
        await this.sendEmail(context);
        channels.push("email");
      } catch (err) {
        errors.push(`email: ${(err as Error).message}`);
      }
    }

    if (errors.length > 0) {
      console.warn("[NotificationService] partial failures:", errors.join(", "));
    }

    return { channels, success: channels.length > 0 || errors.length === 0 };
  }

  private async sendSlack(context: AgentContext): Promise<void> {
    const matched = context.steps.matched;
    const text =
      `*[Smaas] 워크플로우 완료*\n` +
      `업무: ${context.input.title}\n` +
      `담당자 ID: ${matched?.topCandidateId ?? "미배정"}\n` +
      `상태: ${context.status}`;

    await this.postJson(config.slack.webhookUrl, { text });
  }

  private async sendEmail(_context: AgentContext): Promise<void> {
    // SMTP 전송은 nodemailer 등 라이브러리 연동 후 구현
    console.log("[NotificationService] email 전송은 nodemailer 연동 후 활성화됩니다.");
  }

  private postJson(url: string, body: object): Promise<void> {
    return new Promise((resolve, reject) => {
      const data = JSON.stringify(body);
      const parsed = new URL(url);
      const req = https.request(
        {
          hostname: parsed.hostname,
          path: parsed.pathname + parsed.search,
          method: "POST",
          headers: { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(data) },
        },
        (res) => {
          if (res.statusCode && res.statusCode >= 400) {
            reject(new Error(`HTTP ${res.statusCode}`));
          } else {
            resolve();
          }
        }
      );
      req.on("error", reject);
      req.write(data);
      req.end();
    });
  }
}
